### PR TITLE
Constraints for image captioning module

### PR DESCRIPTION
**Pull request by lorisbaz**
Adding
- ~~Constrains input from plain file.~~
- Image Captioning now supports constrained decoding.
- Image Captioning: zero padding of features now allows input features of different shape for each image.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,12 @@ Note that Sockeye has checks in place to not translate with an old model that wa
 
 Each version section may have have subsections for: _Added_, _Changed_, _Removed_, _Deprecated_, and _Fixed_.
 
+## [1.18.34]
+### Added
+- Constrains input from plain file.
```
Comment (fhieber) [inline CHANGELOG.md]: This description isn't clear to me: what constrains the input?

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,12 @@ Note that Sockeye has checks in place to not translate with an old model that wa
 
 Each version section may have have subsections for: _Added_, _Changed_, _Removed_, _Deprecated_, and _Fixed_.
 
+## [1.18.34]
+### Added
+- Constrains input from plain file. 
+- Contraints can be used for image captioning module. 
+- Zero padding of features enables to have input features of different shape for each image.
```
Comment (fhieber) [inline CHANGELOG.md]: Could you rephrase to make clear that this refers to image captioning? Something like: "Image Captioning: zero padding of features now allows input features of different shape for each image."

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,12 @@ Note that Sockeye has checks in place to not translate with an old model that wa
 
 Each version section may have have subsections for: _Added_, _Changed_, _Removed_, _Deprecated_, and _Fixed_.
 
+## [1.18.34]
+### Added
+- Constrains input from plain file. 
+- Contraints can be used for image captioning module.
```
Comment (fhieber) [inline CHANGELOG.md]: maybe: Image Captioning now supports constrained decoding?

File: sockeye/arguments.py [hunk 4]
```diff
@@ -1079,6 +1079,12 @@ def add_inference_args(params):
                                     "Optionally, a list of factors can be provided: "
                                     "{'text': 'some input string', 'factors': ['C C C', 'X X X']}.")
 
+    decode_params.add_argument('--input-constraints',
```
Comment (fhieber) [inline sockeye/arguments.py]: afaik, this introduces an inconsistency with input from a file or stdin. I think we intentionally restricted constrained decoding to json input to avoid parsing multiple files and having to add tons of checks whether they have the same number of lines as the input file / stdin.

File: sockeye/image_captioning/captioner.py [hunk 5]
```diff
@@ -109,6 +109,10 @@ def main():
     params = arguments.ConfigArgumentParser(description='Image Captioning CLI')
     arguments_image.add_image_caption_cli_args(params)
     args = params.parse_args()
+    run_captining(args)
```
Comment (fhieber) [inline sockeye/image_captioning/captioner.py]: typo: captioning

File: sockeye/image_captioning/utils.py [hunk 6]
```diff
@@ -164,3 +164,33 @@ def save_feature(path: str,
         np.save(path, data)
         path += ".npy"
     return path
+
+
+def zero_pad_features(feat: List[np.ndarray],
```
Comment (fhieber) [inline sockeye/image_captioning/utils.py]: I'd spell out variable names, i.e. 'features'

File: sockeye/image_captioning/utils.py [hunk 7]
```diff
@@ -164,3 +164,33 @@ def save_feature(path: str,
...
+                      tshape: tuple) -> List[np.ndarray]:
```
Comment (fhieber) [inline sockeye/image_captioning/utils.py]: same here: target_shape?

File: sockeye/image_captioning/utils.py [hunk 8]
```diff
@@ -164,3 +164,33 @@ def save_feature(path: str,
...
+    """
+    Zero pad to numpy array.
+
+    :param feat: List of numpy arrays.
+    :param tshape: Target shape of each numpy array in the list feat. Note:
+                   tshape should be greater that the largest shapes in feat.
+    :return: A list of padded numpy arrays.
+    """
+    pad_feat = []
+    for f in feat:
```
Comment (fhieber) [inline sockeye/image_captioning/utils.py]: f -> feature

File: sockeye/image_captioning/utils.py [hunk 9]
```diff
@@ -164,3 +164,33 @@ def save_feature(path: str,
...
+        fshape = f.shape
+        if len(fshape)<len(tshape):  # add extra dimensions
```
Comment (fhieber) [inline sockeye/image_captioning/utils.py]: spacing

File: sockeye/image_captioning/utils.py [hunk 10]
```diff
@@ -164,3 +164,33 @@ def save_feature(path: str,
...
+            for i in range(len(tshape)-len(fshape)):
+                f = np.expand_dims(f, axis=len(f.shape)+1)
+            fshape = f.shape
+        elif len(fshape)>len(tshape):
+            raise ValueError("Provided target shape must be bigger then the original "
+                             "shape. (provided: {}, original {})".format(len(tshape), len(fshape)))
+        diff_shape = np.subtract(tshape, fshape)
+        if np.any(diff_shape<0):
```
Comment (fhieber) [inline sockeye/image_captioning/utils.py]: pep8: spacing

File: sockeye/image_captioning/utils.py [hunk 11]
```diff
@@ -164,3 +164,33 @@ def save_feature(path: str,
...
+            raise ValueError("Provided target values must be bigger then the original "
+                             "values for each dimension. (provided: {}, original {})".format(tshape, fshape))
+        diff_shape = [[0, d] for d in diff_shape]  # pad format: ((before_1, after_1), ... (before_N, after_N))
+        p = np.pad(f, diff_shape, 'constant', constant_values=0)
+        pad_feat.append(p)
+    return pad_feat
```
Comment (fhieber) [inline sockeye/image_captioning/utils.py]: missing newline

File: sockeye/inference.py [hunk 12]
```diff
@@ -754,24 +754,30 @@ def make_input_from_factored_string(sentence_id: int,
     return TranslatorInput(sentence_id=sentence_id, tokens=tokens, factors=factors)
 
 
-def make_input_from_multiple_strings(sentence_id: int, strings: List[str]) -> TranslatorInput:
+def make_input_from_multiple_strings(sentence_id: int,
+                                     strings: List[str],
+                                     constraints: str = None) -> TranslatorInput:
```
Comment (fhieber) [inline sockeye/inference.py]: type: Optional[str] if it can be None

File: sockeye/inference.py [hunk 13]
```diff
@@ -754,24 +754,30 @@ def make_input_from_factored_string(sentence_id: int,
...
     """
     Returns a TranslatorInput object from multiple strings, where the first element corresponds to the surface tokens
     and the remaining elements to additional factors. All strings must parse into token sequences of the same length.
 
     :param sentence_id: An integer id.
     :param strings: A list of strings representing a factored input sequence.
+    :param constraints: A string with constraints.
```
Comment (fhieber) [inline sockeye/inference.py]: "Optional string with constraints" ?

File: sockeye/translate.py [hunk 14]
```diff
@@ -166,6 +174,7 @@ def read_and_translate(translator: inference.Translator,
     :param input_file: Optional path to file which will be translated line-by-line if included, if none use stdin.
     :param input_factors: Optional list of paths to files that contain source factors.
     :param input_is_json: Whether the input is in json format.
+    :param input_constraints: Optional path to file which will contain constrains for each source sentence.
```
Comment (fhieber) [inline sockeye/translate.py]: typo: constraints

File: CHANGELOG.md (hunk 1 above)
Comment (lorisbaz) [reply CHANGELOG.md]: I can update that description. The only way to input constraints was via json file. 

Now you can use a plain file. In each line, you have a list of space-separated strings (constraints) which correspond to an image (or source sentence). This is a parallel file to the one specified in --input

File: CHANGELOG.md (hunk 2 above)
Comment (lorisbaz) [reply CHANGELOG.md]: Ok.

File: CHANGELOG.md (hunk 3 above)
Comment (lorisbaz) [reply CHANGELOG.md]: Ok.

File: sockeye/arguments.py (hunk 4 above)
Comment (lorisbaz) [reply sockeye/arguments.py]: To me it makes sense to be flexible here. You can still use the old way with json.

File: sockeye/image_captioning/captioner.py (hunk 5 above)
Comment (lorisbaz) [reply sockeye/image_captioning/captioner.py]: Ops, I'll correct it

File: sockeye/image_captioning/utils.py (hunk 6 above)
Comment (lorisbaz) [reply sockeye/image_captioning/utils.py]: Ok.

File: sockeye/image_captioning/utils.py (hunk 7 above)
Comment (lorisbaz) [reply sockeye/image_captioning/utils.py]: Ok.

File: sockeye/image_captioning/utils.py (hunk 8 above)
Comment (lorisbaz) [reply sockeye/image_captioning/utils.py]: Ok.

File: sockeye/image_captioning/utils.py (hunk 9 above)
Comment (lorisbaz) [reply sockeye/image_captioning/utils.py]: Ok.

File: sockeye/image_captioning/utils.py (hunk 10 above)
Comment (lorisbaz) [reply sockeye/image_captioning/utils.py]: I know. For some reason, my brain wants compact lines 😄

File: sockeye/image_captioning/utils.py (hunk 11 above)
Comment (lorisbaz) [reply sockeye/image_captioning/utils.py]: 🆗

File: sockeye/inference.py (hunk 12 above)
Comment (lorisbaz) [reply sockeye/inference.py]: Right.

File: sockeye/inference.py (hunk 13 above)
Comment (lorisbaz) [reply sockeye/inference.py]: Sounds good.

File: sockeye/translate.py (hunk 14 above)
Comment (lorisbaz) [reply sockeye/translate.py]: 🆗

File: CHANGELOG.md (hunk 1 above)
Comment (fhieber) [reply CHANGELOG.md]: so are these then only single-word constraints? How about several multi-word constraints (which is supported by JSON)?

Comment (lorisbaz) [reply CHANGELOG.md]: Oh I see. How about having them separated by some special character (e.g., comma or tab)?

Comment (lorisbaz) [reply CHANGELOG.md]: E.g.: blue sky, cow, grass, ...

Comment (fhieber) [reply CHANGELOG.md]: Then we would need to rewrite the entire parsing logic of separating tokens. You can see how complicated this already is with the factored strings. In my opinion, inputs with constraints and/or factors is inherently structured data and not just plain strings. JSON provides a good format for such use cases, whereas plain file parsing is quite error-prone. 

I don't want to enforce my opinion on this though. Maybe others can chime in on this discussion? We could also separate the constrained decoding for image captioning feature from this API change to unblock this PR.

Comment (fhieber) [reply CHANGELOG.md]: You mean "source sentence", right?

Creating different constraints for an input sentence is already easy to do. You can use the sockeye.lexical_constraints CLI to produce inputs with arbitrary constraints and pass/pipe them directly to sockeye.translate / sockeye.captioner with the `--json-input` flag.

I think we should hold off on changing the overall API / behavior of the API/translate CLI with this PR. This PR should concentrate on enabling constraints for image captioning.

Comment (lorisbaz) [reply CHANGELOG.md]: Yes. I meant source sentence.

lexical_constraints expects a sentence as input + constraints. Would it work with the image relative path?

I wanted to avoid to use it, because in my case I have to first generate the constraints, then the file with source sentence (image path) + constraints, then apply lexical_constraints and now I can perform inference. It is a bit tricky sequence of steps.

I can revert the changes and remove the option ``--input-constraints``, even though I think it might be useful.

Comment (fhieber) [reply CHANGELOG.md]: I don't know how your captioner currently handles the content of the 'text' field in the JSON object. If it uses it as a relative image path this should already work.

In the end its up to you how you define the API for the image captioning use case, but I vote for keeping the general translation API (sockeye.translate) as is by reverting the changes in the files out side of `sockeye/image_captioning/`.

**Comment by lorisbaz** [comment]
@fhieber I reverted the ``--input-constraints`` option and covered the other small issues listed above.

File: sockeye/inference.py [hunk 15]
```diff
@@ -754,7 +754,8 @@ def make_input_from_factored_string(sentence_id: int,
     return TranslatorInput(sentence_id=sentence_id, tokens=tokens, factors=factors)
 
 
-def make_input_from_multiple_strings(sentence_id: int, strings: List[str]) -> TranslatorInput:
+def make_input_from_multiple_strings(sentence_id: int,
```
Comment (fhieber) [inline sockeye/inference.py]: unnecessary change

File: test/unit/test_inference.py [hunk 16]
```diff
@@ -311,7 +311,7 @@ def test_failed_make_input_from_valid_json_string(text, text_key, factors, facto
 @pytest.mark.parametrize("strings",
                          [
                              ["a b c"],
-                             ["a b c", "f1 f2 f3", "f3 f3 f3"]
+                             ["a b c", "f1 f2 f3", "f3 f3 f3"],
```
Comment (fhieber) [inline test/unit/test_inference.py]: probably not an issue, but adding the comma could be reverted.

File: sockeye/image_captioning/utils.py [hunk 17]
```diff
@@ -164,3 +164,33 @@ def save_feature(path: str,
...
+    :param target_shape: Target shape of each numpy array in the list feat. Note:
+                   target_shape should be greater that the largest shapes in feat.
+    :return: A list of padded numpy arrays.
+    """
+    pad_features = []
+    for feature in features:
+        feature_shape = feature.shape
+        if len(feature_shape) < len(target_shape):  # add extra dimensions
+            for i in range(len(target_shape)-len(feature_shape)):
+                feature = np.expand_dims(feature, axis=len(feature.shape)+1)
```
Comment (fhieber) [inline sockeye/image_captioning/utils.py]: tiny nit: spacing :D

File: sockeye/inference.py (hunk 15 above)
Comment (lorisbaz) [reply sockeye/inference.py]: Agree

File: test/unit/test_inference.py (hunk 16 above)
Comment (lorisbaz) [reply test/unit/test_inference.py]: It was just because I remove the constraints manually. Fixing it.

File: sockeye/image_captioning/utils.py (hunk 17 above)
Comment (lorisbaz) [reply sockeye/image_captioning/utils.py]: Arggggg ☠️ 😄

**Comment by fhieber** [comment]
I won't forget :)

**Comment by mjpost** [comment]
Looks good to me. Two thoughts:

- We could consider using a different keyword other than "tokens" in `TranslatorInput` for the path to the image file. There's no need to overload on this and it could be clearer otherwise.
- Can you add notes to the tutorials to list this feature, for both [lexical constraints](https://github.com/awslabs/sockeye/blob/master/tutorials/constraints/README.md) and [image captioning](https://github.com/awslabs/sockeye/blob/master/tutorials/image_captioning/README.md)?

**Comment by mjpost** [comment]
As an additional thought, note that the [lexical constraints documentation](https://github.com/awslabs/sockeye/blob/master/tutorials/constraints/README.md) provides a command-line module for generating the JSON object from tab-delimited data on STDIN of the form:

    input sentence TAB constraint1 TAB another constraint TAB yet another constraint

If you ignored my thought #1, this would work transparently with image captioning (with field 1 as the path).

**Comment by lorisbaz** [comment]
I removed the option to have a plain file. We will use JSON files as you and Felix suggested.

I will add some notes about lexical constraints in the image captioning tutorial.

**Comment by lorisbaz** [comment]
Hey @mjpost, I added the instructions. Can you have a look and approve if it is ok?
Thanks.

**Comment by mjpost** [comment]
LGTM. You'll have to remerge master once more since we just fixed the pylint issues that caused Travis to fail.

**Comment by lorisbaz** [comment]
Done
